easybuild/framework/easyconfig/easyconfig.py

```diff
@@ -595,8 +596,17 @@ def update(self, key, value, allow_duplicate=True):
             for item in lval:
                 if allow_duplicate or item not in param_value:
                     param_value = param_value + [item]
+        elif isinstance(param_value, tuple):
+            param_value = list(param_value)
+            for item in lval:
+                if allow_duplicate or item not in param_value:
+                    param_value.append(item)
+            param_value = tuple(param_value)
+        elif isinstance(param_value, dict):
+            param_value.update(lval)
```

Comment:
Two things here:

* You shouldn't do an in-place update of `param_value` here, just like we avoid above in the `list` case, but you should take a copy and update that instead. Below we replace the actual easyconfig parameter value (`self[key]`) with the updated `param_value`.

* You shouldn't assume `lval` is a `dict` here, just because `param_value` is. What if `lval` is a `list` of `tuple`?
   In that case we can't actually do the update (since we don't have key-value pairs.
   Similarly above: what if `param_value` is a `list`, but `lval` is a `dict`?

Reply:
I thought param_value was already a local copy, made with the intent of doing in-place updates before pushing it back onto the original?

I see your point about handling all of the crossed cases where the user's input isn't necessarily the same as the target type (lval).

Reply:
See also above.  I am going to change the name, because (at least to me, coming from C++) the name "lval" is really misleading; to me, it implies "left-hand value".  I think it's supposed to mean "*l*ocal *val*ue" or something.  I'd rather call it "inval" ("*in*put *val*ue") or something.